### PR TITLE
New API

### PR DESCRIPTION
**Pull request by omid**
Outcome based on #55 discussion

Remaining tasks:
- [x] Try to make solve the following comment: https://github.com/fmeringdal/rust-rrule/pull/60#discussion_r872610286
- [x] Make `RRule` and `RRuleSet` fields private
- [x] Update Docs
- [x] Update CHANGELOG.md

### CODE REVIEW

**Comment by omid** [comment]
@fmeringdal sry, but I have some homework for you for the weekend 🙈

File: rrule/src/iter/rruleset_iter.rs [hunk 1]
```diff
@@ -70,32 +75,33 @@ impl<'a> RRuleSetIter<'a> {
...
-            None => true,
+            None => false,
             Some(date) => {
-                let dt = date.timestamp();
+                let ts = date.timestamp();
 
                 if !exrules.is_empty() {
-                    let after = date.timezone().timestamp(dt - 1, 0);
-                    let before = date.timezone().timestamp(dt + 1, 0);
+                    let start = date.timezone().timestamp(ts - 1, 0);
```
Comment (omid) [inline rrule/src/iter/rruleset_iter.rs:90]: I don't know what we are doing here exactly.

File: rrule/src/core/date_filter.rs [hunk 2]
```diff
@@ -1,17 +1,17 @@
...
-    &'a Self: 'a + Sized + IntoIterator<IntoIter = T>,
-    T: Iterator<Item = DateTime> + WithError,
+    Self: Sized + Iterator<Item = DateTime> + WithError,
 {
     /// Returns all the recurrences of the rrule.
     ///
     /// Limit must be set in order to prevent infinite loops.
     /// The max limit is `65535`. If you need more please use `into_iter` directly.
-    fn all(&'a self, limit: u16) -> Result<Vec<DateTime>, RRuleError> {
-        super::collect_or_error(self.into_iter(), &None, &None, true, limit)
```
Comment (omid) [inline rrule/src/core/date_filter.rs:14]: I couldn't make the old way work, so users must use `rrule.into_iter().all(...)`.
Any suggestion to revert this `into_iter` thingy change is very welcome.

**Comment by omid** [comment]
We need to reread all the documentations before merge.

**Comment by fmeringdal** [comment]
Oh, that was quick! This is a change I want to think more about before merging. I will try to give som feedback over the next week.

File: rrule/src/core/date_filter.rs (hunk 2 above)
Comment (omid) [reply rrule/src/core/date_filter.rs:14]: If we couldn't revert it back, we can also remove the `IntoIteratorWithCtx` trait.

Comment (omid) [reply rrule/src/core/date_filter.rs:14]: @fmeringdal 
I realized, since it's not possible to iterate over `RRule`. And `RRuleIter` will not be public.
We don't need the `DateFilter` trait at all!

I removed it locally and moved the methods to `RRuleSet`, they work and are happy!
I'll push my changes within a new commit. We can revert if you disagree.

We internally only require `all_between` for `RRuleIter` to calculate exrules. We might be able to even remove that!

**Review by fmeringdal — COMMENTED** [review]
First round of feedback! ;)

File: rrule/src/core/date_filter.rs (hunk 2 above)
Comment (fmeringdal) [reply rrule/src/core/date_filter.rs:14]: Agreed. I was also thinking that `DateFilter` could be removed if `RRule` does not implement it. It also makes the library easier to use as the user does not have to import `DateFilter` to call the methods of `RRuleSet`.

File: rrule/src/iter/rruleset_iter.rs (hunk 1 above)
Comment (fmeringdal) [reply rrule/src/iter/rruleset_iter.rs:90]: The function `is_generated_date_excluded` is called with a date-time that was generated from an `RRULE` or `RDATE` and checks if it matches against an `EXDATE` or `EXRULE`, in which case the date-time will be discarded. The before and after timestamps are used as boundaries when checking the `EXRULE:` iterator for a match against the given date-time so that we don't have to iterate through every occurrence in the `EXRULE` when we know that it has to be in the interval <ts - 1, ts + 1>.

Comment (fmeringdal) [reply rrule/src/iter/rruleset_iter.rs:90]: While you are here, do you mind cleaning up the return code for the function? :)
```rust
if exdates.contains(&ts) {
  return true;
}

false
```

To
 
```rust
exdates.contains(&ts)
```

File: rrule/src/iter/mod.rs [hunk 3]
```diff
@@ -470,10 +454,17 @@ fn make_timeset(
...
-        &properties.freq,
+        &rrule.freq,
         counter_date.hour() as u8,
         counter_date.minute() as u8,
         counter_date.second() as u8,
         counter_date.timestamp_subsec_millis() as u16,
     )
 }
+
+pub trait IntoIteratorWithCtx {
```
[the remark this answers is not in the document]
Comment (omid) [reply rrule/src/iter/mod.rs:465]: You are right. No need :) 
It was needed if we needed to have DateFilter, without that we don't need it!

File: rrule/src/core/rruleset.rs [hunk 4]
```diff
@@ -26,6 +37,13 @@ impl Default for RRuleSet {
...
 
 impl RRuleSet {
+    pub fn new(dt_start: DateTime) -> Self {
+        Self {
+            dt_start,
+            ..Default::default()
+        }
+    }
+
     pub fn rrule(&mut self, rrule: RRule) {
```
[the remark this answers is not in the document]
Comment (omid) [reply rrule/src/core/rruleset.rs:47]: You are right, nice catch, fields shouldn't be public.
But I think method are correct atm. You can change `Unvalidated` one, but there is no method on the `Validated`. So it's immutable!

[the remark this answers is not in the document]
Comment (omid) [reply rrule/src/core/rruleset.rs:47]: The hard part is that I have to update all the tests :))
But np, it will take more time :P

File: rrule/src/core/rruleset.rs [hunk 5]
```diff
@@ -1,15 +1,26 @@
...
 
-#[derive(Debug, Clone)]
+/// A validated Recurrence Rule that can be used to create an iterator.
+#[cfg_attr(feature = "serde", serde_as)]
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(DeserializeFromStr, SerializeDisplay))]
 pub struct RRuleSet {
+    /// The properties specified by this rule.
     pub rrule: Vec<RRule>,
     pub rdate: Vec<DateTime>,
```
[the remark this answers is not in the document]
Comment (omid) [reply rrule/src/core/rruleset.rs:20]: I agree, but then we need to have more methods, to remove, add one (we have this one) and set all and get all for all 4 fields.

File: rrule/src/iter/rruleset_iter.rs (hunk 1 above)
Comment (omid) [reply rrule/src/iter/rruleset_iter.rs:90]: Strange that clippy didn't catch this! 🤔

Comment (omid) [reply rrule/src/iter/rruleset_iter.rs:90]: Mmm, I still don't get the logic there.
And I found some infinite loops there, I think.
I put it in my future PR:

```
#[test]
fn edge_case_1() {
    let rrule: RRuleSet = "DTSTART;TZID=Europe/Berlin:20210101T000000;
RRULE:FREQ=MONTHLY
EXRULE:FREQ=MONTHLY"
        .parse()
        .expect("The RRule is not valid");

    rrule.into_iter().next();
}
```

File: rrule/src/core/rruleset.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (fmeringdal) [reply rrule/src/core/rruleset.rs:47]: > You can change Unvalidated one, but there is no method on the Validated.

Can you elaborate on why this means it is better to make the fields private rather than changing the method signature to take `RRule<Unvalidated>`.

A benefit of accepting an unvalidated `RRule` is that the user don't have to call the `RRule::validate` method, to call `RRule::validate` they also need the start date which is a field on `RRuleSet` so it makes sense that the validation happens inside `RRuleSet`.

Changing all tests sound like a daunting task so let's be sure about the change before actually doing it :D

[the remark this answers is not in the document]
Comment (omid) [reply rrule/src/core/rruleset.rs:47]: I think you mixed two comments, or I misunderstood the first comment in this thread. There is [another comment](https://github.com/fmeringdal/rust-rrule/pull/60#discussion_r878353081) for `RRuleSet` and this one is for `RRule` I think!

Anyway, I agree to make both `RRule` and `RRuleSet` properties, private (or maybe `pub(crate)`).

[the remark this answers is not in the document]
Comment (omid) [reply rrule/src/core/rruleset.rs:47]: > Can you elaborate on why this means it is better to make the fields private rather than changing the method signature to take RRule<Unvalidated>?

And here, I meant, when `RRule` fields become private, you cannot make a `Validated` invalid `RRule`. So it will be safe.

**Comment by omid** [comment]
@fmeringdal to avoid hundreds of conflicts, I had to merge my other PR into this. All changes are in [one commit](https://github.com/fmeringdal/rust-rrule/pull/60/commits/00575af543a9bcf626abd810c1e2bdc1616149b5). Mostly strict clippy checks + some documentation + renaming `properties` from the code to `rrule` + adding some edge case tests (inc. two ignores found failing test cases)

File: rrule/src/core/rruleset.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (omid) [reply rrule/src/core/rruleset.rs]: I [made](https://github.com/fmeringdal/rust-rrule/pull/60/commits/22db507391c2e7c31a35a5db000987f44da4372c) all `RRule` and `RRuleSet` fields private (pub(crate)).
Now the code you pasted here shouldn't work and the API should be safe, yes?

[the remark this answers is not in the document]
Comment (omid) [reply rrule/src/core/rruleset.rs]: And to fix the `could not compile rrule due to 2771 previous errors` issue, I moved the tests inside.

File: rrule/src/tests/edge_cases.rs [hunk 6]
```diff
@@ -0,0 +1,38 @@
+use crate::RRuleSet;
+
+#[test]
+fn non_ascii_string() {
```
Comment (fmeringdal) [inline rrule/src/tests/edge_cases.rs:4]: Do you think thes tests in this file could be added to `rrule/tests/regressions.rs`?

**Review by fmeringdal — CHANGES_REQUESTED** [review]
Great work, thank you! This is close to getting merged.
One thing for future PRs: I think it would be better to only focus on one change per PR if possible, then it is easier to review and we will be able to merge it faster ;)

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -17,7 +17,10 @@ jobs:
       - name: Linting
         run: |
           cd rrule
-          cargo clippy -- -D warnings
+          cargo clippy
```
[the remark this answers is not in the document]
Comment (omid) [reply .github/workflows/ci.yml:20]: We shouldn't allow some warnings which should stay as a warning :D For example, we have some long functions which we must split, but if we allow them, then we will ignore them forever!

I would say, warning should stay as warning, they shouldn't block us from continuing, but we should see the message until fixing them

File: rrule/src/core/datetime.rs [hunk 8]
```diff
@@ -21,28 +21,34 @@ impl Time {
...
+    pub fn time(self) -> u64 {
+        (u64::from(self.hour) * 60 * 60 + u64::from(self.minute) * 60 + u64::from(self.second))
+            * 1000
+            + u64::from(self.millisecond)
     }
 
     pub fn to_naive_time(self) -> NaiveTime {
-        NaiveTime::from_hms(self.hour as u32, self.minute as u32, self.second as u32)
+        NaiveTime::from_hms(
+            u32::from(self.hour),
```
[the remark this answers is not in the document]
Comment (omid) [reply rrule/src/core/datetime.rs:32]: That is the automatic clippy changes with `--fix`, and the reason is this: https://rust-lang.github.io/rust-clippy/master/index.html#cast_lossless

[the remark this answers is not in the document]
Comment (fmeringdal) [reply rrule/src/core/datetime.rs:32]: Ah, thanks!

File: .github/workflows/ci.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (fmeringdal) [reply .github/workflows/ci.yml:20]: Hmm, you have a point. There is still a lot of refactorings needed in the library. Could your leave a TODO comment there for my sanity? :D Something like: `TODO: Ideally we would like to do deny warnings as well in the future when the code quality has improved`

**Comment by omid** [comment]
@fmeringdal a kind reminder after 2 weeks :)

**Comment by fmeringdal** [comment]
I didn't forget this, just noticed that you hadn't marked the "Update CHANGELOG.md" as done until today, so I assumed you wanted to write some more :)

I think this PR is good to go, but want to keep it open for a couple more days to think if there is anything more that needs to be done. One follow up issue after this PR is to go through all of the different number types we use and see if they can be made more compatible so that we can avoid the enormous amount of casts we do atm.

**Comment by omid** [comment]
@fmeringdal Oh, sorry, no, I couldn't find anything else for the changelog. I would like to ask you, since you might be a heavy user of your library, if you find something missing in the changelog, please add it.

Regarding the types, I agree. AFAICR, I could eliminate type casts related to one struct only.

**Comment by omid** [comment]
In the meanwhile this PR was open, `clap`, `clap_lex` and some other packages change the MSRV in a patch or minor version :(
So we can go to an older version of `clap` (and maybe some other packages) or upgrade the MSRV, to the first version of edition 2021. @fmeringdal wdyt?

**Comment by fmeringdal** [comment]
@omid I created two follow up issues: #62 and #63.
Is there anything more you would like to do in this PR? Or any follow up work you would like to create?

**Comment by omid** [comment]
@fmeringdal Nope, thanks

**Comment by fmeringdal** [comment]
🎉 This is a big leap forwards to v1.0! Thanks @omid . v0.8 will be released soon
